### PR TITLE
WIP: ValueMappings: improve config UI and allow string inputs

### PR DESCRIPTION
**Pull request by ryantxu**
The current value mapping UI is difficult when doing anything complex:
* ValueMap should support strings (currently just numbers)
* hard to understand what is mapped when many items are entered
* rules should be sortable since they are evaluated in their sort order

image

For reference, things used to look kinda like this:
image

TODO:

- [x] Actually update the models
- [ ] React.FC vs PureComponent?
- [x] remove operator?
- [ ] id?  is that just the index?
- [ ] grid/flex sizing vs absolute?
- [ ] fix tests

### CODE REVIEW

**Comment by dprokop** [comment]
String values mapping will be fixed with https://github.com/grafana/grafana/pull/25929

File: packages/grafana-data/src/types/valueMapping.ts [hunk 1]
```diff
@@ -4,9 +4,9 @@ export enum MappingType {
 }
 
 interface BaseMap {
-  id: number;
-  operator: string;
-  text: string;
+  id: number; // Could/should be derived from the array index!!!
+  // operator: string; // ???????????????
```
Comment (dprokop) [inline packages/grafana-data/src/types/valueMapping.ts:8]: If TS does not complain then I think we can remove this

File: packages/grafana-ui/src/components/ValueMappingsEditor/RangeMappingRow.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  dragClass: string;
+  mapping: RangeMap;
+  index: number;
+  onChange: (index: number, mapping: ValueMapping) => void;
+}
+
+export class RangeMapRow extends PureComponent<RangeMapProps> {
+  constructor(props: RangeMapProps) {
+    super(props);
+  }
```
Comment (dprokop) [inline packages/grafana-ui/src/components/ValueMappingsEditor/RangeMappingRow.tsx:15]: This kind of constructors are not necessary :)

File: packages/grafana-ui/src/components/ValueMappingsEditor/ValueMappingRow.tsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  dragClass: string;
+  mapping: ValueMap;
+  index: number;
+  onChange: (index: number, mapping: ValueMapping) => void;
+}
+
+export class ValueMapRow extends PureComponent<ValueMapProps> {
+  constructor(props: ValueMapProps) {
+    super(props);
+  }
```
Comment (dprokop) [inline packages/grafana-ui/src/components/ValueMappingsEditor/ValueMappingRow.tsx:15]:
```suggestion
```

**Comment by torkelo** [comment]
This needs another UI update. This is just too inconsistent from the rest of the sidebar introducing a new mix of UI patterns, not following normal inline forms style nor the new forms. 

It's tricky, do not know what the solution is given the width constraints. But do not like the mix of inline forms labels and input prefixes (that causes misalignment). 

Screenshot from 2020-07-01 10-02-46

One idea is to use the collapsable row (that we use in transforms and query editors) to show a collapsed form of the mapping and expand it to show a full form

**Comment by dprokop** [comment]
@torkelo I agree. I like the idea with the collapsible row too, gonna experiment with this.

File: packages/grafana-ui/src/components/ValueMappingsEditor/DraggableMappingRow.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        </div>
+        {children}
+      </>
+    );
+  }
+);
+const getDraggableRowStyles = (theme?: GrafanaTheme) => {
+  if (!theme) {
+    theme = {} as GrafanaTheme;
+  }
```
Comment (ryantxu) [inline packages/grafana-ui/src/components/ValueMappingsEditor/DraggableMappingRow.tsx:28]: @dprokop -- obviously something funky here.  this was required for typings

`useTheme()` makes an optional theme?

**Comment by ryantxu** [comment]
Collapsible is interesting -- it would require many clicks when trying to make a long list consistent though.  I was messing with trying to get flexbox to stack when there is not enough space to show in a single row... 

image

File: packages/grafana-ui/src/components/ValueMappingsEditor/DraggableMappingRow.tsx (hunk 4 above)
Comment (dprokop) [reply packages/grafana-ui/src/components/ValueMappingsEditor/DraggableMappingRow.tsx:28]: uhm, must have missed that. No need for that at all, useTheme assures theme object.

**Pull request closed without merging**